torchao/sparsity/prototype/training/pointwise_ops.py

```diff
@@ -0,0 +1,87 @@
+from functools import partial
+import torch
+from torch.sparse import SparseSemiStructuredTensor 
+from torchao.sparsity.prototype.training.autograd import semi_sparse_sparsify_like
+
+def _semi_sparse_pointwise_op(func, types, args=(), kwargs=None, sparsify_like_args_list=()):
+    """
+    adds pointwise op support for semi-structured tensors. 
+
+    Assumes that at least one of the arguments in arg is a SparseSemiStructuredTensor.
+    The last instance of a SparseSemiStructuredTensor is used as the reference mask to sparsify the others tensors passed in args. 
+    sparsify_like_args_list is used to specify which arguments to sparsify like the reference tensor.
+    """
+    reference_sparse_tensor = None
+    for tensor in args:
+        if isinstance(tensor, SparseSemiStructuredTensor):
+            reference_sparse_tensor = tensor
+    assert reference_sparse_tensor is not None
+
+    def handle_arg(i, tensor):
+        if isinstance(tensor, torch.Tensor):
+            if not isinstance(tensor, SparseSemiStructuredTensor):
+                if i in sparsify_like_args_list:
+                    tensor = semi_sparse_sparsify_like(tensor, reference_sparse_tensor)
+                else:
+                    raise ValueError(
+                        f"Operation {func.__module__}.{func.__name__} on {type(reference_sparse_tensor)} requires all operands to "
+                        f"be {type(reference_sparse_tensor)}, but operand {i} is a {type(tensor)}"
+                    )
+            else:
+                if (
+                    tensor.compressed_swizzled_bitmask is None
+                    or reference_sparse_tensor.compressed_swizzled_bitmask is None
+                    or tensor.compressed_swizzled_bitmask.data_ptr() != reference_sparse_tensor.compressed_swizzled_bitmask.data_ptr()
+                    or tensor.compressed_swizzled_bitmask.stride() != reference_sparse_tensor.compressed_swizzled_bitmask.stride()
+                ):
+                    raise ValueError(
+                        f"Operation {func.__module__}.{func.__name__} on {type(reference_sparse_tensor)} requires all operands to be "
+                        f"{type(reference_sparse_tensor)} with the same sparsity pattern"
+                    )
+        return tensor
+
+    args_updated = [ handle_arg(i, tensor) for i, tensor in enumerate(args) ]
+
+    return reference_sparse_tensor.__class__(
+        reference_sparse_tensor.shape,
+        func(*[
+                x.packed if isinstance(x, SparseSemiStructuredTensor) else x
+                for x in args_updated
+            ]),
+        reference_sparse_tensor.meta,
+        func(
+            *[
+                x.packed_t if isinstance(x, SparseSemiStructuredTensor)
+                else x for x in args_updated
+            ]
+        ),
+        reference_sparse_tensor.meta_t,
+        reference_sparse_tensor.compressed_swizzled_bitmask,
+    )
+
+# Add pointwise ops to the dispatch table
```

Comment:
how do you decide what to do add here? It seems like you looked at what's most used in actual modeling code but my question would be why not all pointwise ops in pytorch?

Reply:
I just added what was necessary for our experiments, for some pointwise ops, like mul you need to define the sparsification_like_args_list, so you cannot apply naively for all pointwise ops. 

But it would make sense to add support for all the naive ones, I can add in a subsequent PR.